### DISCUSSION TITLE
About CLASSPATH

### DISCUSSION THREAD

**Discussion by AntonioSun** [smicyk/groovy-jmeter#115] [Q&A]
_Originally posted by @smicyk in https://github.com/smicyk/groovy-jmeter/issues/112#issuecomment-1345631874_

>   Please, remember that when you execute the script you need to add the plugin jar to the classpath

This is a follow up question on above, and more about [groovy-jmeter](https://github.com/smicyk/groovy-jmeter)'s internal behaviour -- how it constructs  jmeter script and how it calls jmeter internally.

Suppose that I've manually created from [groovy-jmeter](https://github.com/smicyk/groovy-jmeter)'s groovy  as jmeter script, and I invoke my jmeter on such jmeter script, then for any plugin that jmeter uses, I don't need to specify them on jmeter's command line, right?

Now, when it comes to `groovy  my-groovy-jmeter-script.groovy`, i.e., constructing  jmeter script and calling jmeter all internally by [groovy-jmeter](https://github.com/smicyk/groovy-jmeter), I'd assume that, for any plugin that jmeter uses, I don't need to specify them on jmeter's command line either, even for the [another InfluxDb BackendListener](https://github.com/smicyk/groovy-jmeter/issues/112), right?

**Comment by smicyk** [comment]
Hi,

so it really depends how you execute your scripts but let's go through different scenarios:

1) You use the JMeter GUI to create your script. When you use the plugin you must first download it with Plugin Manager (most likely but you can also put the required stuff in <jmeter home>lib/ext folder manually). Either way you have some .jar file with the plugin code in the lib/ext folder. Then you can see the new test elements, listeners etc in the GUI. When you run the script from UI everything will work since all plugins are on JMeter GUI classpath.

2) You run the script created in JMeter GUI on command line with jmeter.bat, then you don't need to specify anything on the command line since jmeter.bat automatically reads the /lib/ext folder and it should be fine. You don't need to specify anything on command line (assuming you already put the plugins in lib/ext).

3) You generate jmeter script (.jmx) with groovy-jmeter. Here it depends on what type of plugin it is. If you have plugin for backend listener you most likely don't need to specify anything as it will only generate the .jmx and don't need specific classes to serialize them to xml/jmx file. But if the plugin is some kind of controller or sampler then you need to specify the plugin jar file in the classpath. Otherwise the groovy-jmeter would not know how to serialize the new plugin to xml/jmx file.

4) You run script with groovy-jmeter. You always need the plugin .jar file on the classpath.

Saying that we might agree that there should be some possibility to run groovy-jmeter without specifying classpath, but then we would need some default folder where the groovy-jmeter will look at for the plugin .jar files. Same as the JMeter GUI is doing right now.

As an alternative you can try to add required dependencies with @Grab annotation, e.g. something like:

```groovy
@Grab('kg.apc:jmeter-plugins-dummy:0.4')
@GrabExclude('xml-apis:xml-apis')
@GrabExclude('xercesi:xercesImpl')
```

The script from examples/plugins/include would look like this:

```groovy
@GrabConfig(systemClassLoader=true)
@Grab('net.simonix.scripts:groovy-jmeter')
@Grab('kg.apc:jmeter-plugins-dummy:0.4')
@GrabExclude('xml-apis:xml-apis')
@GrabExclude('xercesi:xercesImpl')

@groovy.transform.BaseScript net.simonix.dsl.jmeter.TestScript script

start {
    plan {
        group(users: 10, rampUp: 1) {
            include 'dummy_plugin.jmx'
        }

        summary(file: 'script.jtl', enabled: true)
    }
}
```
But, first the plugin needs to be in public maven repository or you need your own local repository with installed plugins or your company can provide you some kind artifactory where groovy can look for the plugins.

The second you need to deal with some dependencies problems since plugin can use older version of some libraries which is not compatible with Jmeter dependencies. Here I must add those GrabExclude so they don't break the script.
